reports/generator/generator.go

```diff
@@ -0,0 +1,220 @@
+package generator
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+
+	"github.com/iryonetwork/wwm/reports"
+)
+
+type (
+	generator struct {
+		storage reports.Storage
+		writer  ReportWriter
+		logger  zerolog.Logger
+	}
+)
+
+const dataKeyCategory = "/category"
+
+const sourceData = "Data"
+const sourceFileID = "FileID"
+const sourceVersion = "Version"
+const sourcePatientID = "PatientID"
+const sourceCreatedAt = "CreatedAt"
+const sourceUpdatedAt = "UpdatedAt"
+
+// Generate generates report
+func (g *generator) Generate(ctx context.Context, writer ReportWriter, reportSpec ReportSpec, createdAtStart *strfmt.DateTime, createdAtEnd *strfmt.DateTime) error {
+	files, err := g.storage.Find("", map[string]string{dataKeyCategory: reportSpec.FileCategory}, createdAtStart, createdAtEnd)
+	if err != nil {
+		return errors.Wrapf(err, "failed to fetch files")
+	}
+
+	err = writer.Write(reportSpec.Columns)
+	if err != nil {
+		return errors.Wrapf(err, "failed to write report header")
+	}
+
+	for fileIndex, file := range *files {
+		var dataMap map[string]interface{}
+		if err := json.Unmarshal([]byte(file.Data), &dataMap); err != nil {
+			return errors.Wrapf(err, "failed to unmarshal file data")
+		}
+
+		row := []string{}
+		for _, column := range reportSpec.Columns {
+			spec, ok := reportSpec.ColumnsSpecs[column]
+			if !ok {
+				return errors.Errorf("could not find a spec for column '%s'", column)
+			}
+			switch spec.Source {
+			case sourceData:
+				_, value := g.generateValueFromData(spec, &file, &dataMap, "")
+				row = append(row, strings.TrimSpace(value))
+			case sourceFileID:
+				row = append(row, file.FileID)
+			case sourceVersion:
+				row = append(row, file.Version)
+			case sourcePatientID:
+				row = append(row, file.PatientID)
+			case sourceCreatedAt:
+				row = append(row, file.CreatedAt.String())
+			case sourceUpdatedAt:
+				row = append(row, file.UpdatedAt.String())
+			}
+		}
+		err = writer.Write(row)
+		if err != nil {
+			return errors.Wrapf(err, "failed to write report row")
+		}
+	}
+
+	return nil
+}
+
+// Generate generates report
+func (g *generator) GenerateGroupedByPatientID(ctx context.Context, writer ReportWriter, reportSpec ReportSpec, createdAtStart *strfmt.DateTime, createdAtEnd *strfmt.DateTime) error {
+	files, err := g.storage.Find("", map[string]string{dataKeyCategory: reportSpec.FileCategory}, createdAtStart, createdAtEnd)
+	if err != nil {
+		return errors.Wrapf(err, "failed to fetch files")
+	}
+
+	err = writer.Write(reportSpec.Columns)
+	if err != nil {
+		return errors.Wrapf(err, "failed to write report header")
+	}
+
+	var patientIdFileIndexMap map[string][]string
+	for fileIndex, file := range *files {
+		if _, ok := patientIdFileIdMap[file.PatientID]; !ok {
+			patientIdFileIdMap[file.PatientID] = []string{fileIndex}
```

Comment:
cannot use fileIndex (variable of type int) as string value in array or slice literal